### PR TITLE
Fix collapse group on icon click

### LINKED ISSUE #474: cannot collapse expanded group of nodes

### LINKED ISSUE #511: Collapse icon not working in grouped nodes
Steps to reproduce:
- Group some nodes
- Click "collapse" button

Expected behaviour: nodes group will be collapsed 
Behaviour now: nothing happens, there is a stack trace in console:
```
main.js?a8637bffcfddf15d0927:1 Uncaught TypeError: Cannot read property 'type' of undefined
    at e.findNodeDefinitionId (main.js?a8637bffcfddf15d0927:1)
    at e.edgesForNode (main.js?a8637bffcfddf15d0927:1)
    at e._canHaveMoreInputs (main.js?a8637bffcfddf15d0927:1)
    at e.canMakeLink (main.js?a8637bffcfddf15d0927:1)
    at Object.u (main.js?a8637bffcfddf15d0927:1)
    at t.handleInjectBetweenNodes (main.js?a8637bffcfddf15d0927:1)
    at r.<anonymous> (main.js?a8637bffcfddf15d0927:1)
    at g (vendors.js?a8637bffcfddf15d0927:192)
    at p (vendors.js?a8637bffcfddf15d0927:192)
    at c (vendors.js?a8637bffcfddf15d0927:192)
```

### PR DESCRIPTION
**Pull request by mateuszkp96**
This closes #474, closes #511 
image
image

Now we can drop nodes above and below groups:
image
image
image
image

TODO - handle inject of group when is dropped on line

### CODE REVIEW

File: ui/client/components/graph/Graph.js [hunk 1]
```diff
@@ -214,12 +214,26 @@ class Graph extends React.Component {
       const source = JointJsGraphUtils.findCell(this.graph, linkBelowCell.attributes.source.id)
       const target = JointJsGraphUtils.findCell(this.graph, linkBelowCell.attributes.target.id)
       const middleMan = cellView.model
+
+      // TODO: handle inject when group is middleman
+      if (NodeUtils.nodeIsGroup(middleMan.attributes.nodeData))  {
```
Comment (PPiedel) [inline ui/client/components/graph/Graph.js]: Maybe we should notify user that injecting a group is not possible yet?

File: ui/client/components/graph/NodeUtils.js [hunk 2]
```diff
@@ -205,9 +205,15 @@ class NodeUtils {
   }
 
   //TODO: methods below should be based on backend data, e.g. Subprocess can have outputs or not - based on individual subprocess...
-  hasInputs = (node) => (node.type !== "Source" && node.type !== "SubprocessInputDefinition")
+  hasInputs = (node) => {
```
Comment (PPiedel) [inline ui/client/components/graph/NodeUtils.js]: These methods are very similar and could be merged

**Comment by PPiedel** [comment]
Please add description with issue resolved

File: ui/client/components/graph/Graph.js (hunk 1 above)
Comment (lciolecki) [reply ui/client/components/graph/Graph.js]: But show notifiaction type = info, not error :)

**Comment by mateuszkp96** [comment]
Added description with issue resolved

On inject group node: 
image

Info:
image

File: ui/client/components/graph/NodeUtils.js (hunk 2 above)
Comment (mateuszkp96) [reply ui/client/components/graph/NodeUtils.js]: I did a small refactor

File: ui/client/components/graph/Graph.js (hunk 1 above)
Comment (mateuszkp96) [reply ui/client/components/graph/Graph.js]: done

**Comment by witekw** [comment]
image

Graph above - injecting a node between two nodes on expanded group, collapsing. The newly added node is not a member of a group.

File: ui/client/components/graph/Graph.js [hunk 3]
```diff
@@ -214,12 +214,27 @@ class Graph extends React.Component {
...
-      //TODO: consider doing this check in actions.js?
-      if (GraphUtils.canInjectNode(this.props.processToDisplay, source, middleMan, target, this.props.processDefinitionData)) {
+      const middleManNode = middleMan.attributes.nodeData
+
+      const sourceNodeData = source.attributes.nodeData
+      const sourceNode = NodeUtils.nodeIsGroup(sourceNodeData) ? _.last(sourceNodeData.nodes) : sourceNodeData
+
+      const targetNodeData = target.attributes.nodeData
+      const targetNode = NodeUtils.nodeIsGroup(targetNodeData) ? _.head(targetNodeData.nodes) : targetNodeData
+
```
Comment (PPiedel) [inline ui/client/components/graph/Graph.js]: blank line

**Comment by mateuszkp96** [comment]
Now notification appears when trying to inject a node in expanded group
image
image

File: ui/client/components/graph/Graph.js (hunk 3 above)
Comment (mateuszkp96) [reply ui/client/components/graph/Graph.js]: removed

File: ui/client/components/graph/NodeUtils.js [hunk 4]
```diff
@@ -204,10 +204,22 @@ class NodeUtils {
     return `${edge.from}-${edge.to}`
   }
 
+  groupIncludesOneOfNodes = (nodeGroup, nodeIds)  => _.some(nodeGroup.nodes, (node) => _.includes(nodeIds, _.get(node, "id") || node))
+
+  groupIncludesAllOfNodes = (nodeGroup, nodeIds) => _.every(nodeGroup.nodes, (node) => _.includes(nodeIds, _.get(node, "id") || node))
+
+  nodesAreInOneGroup = (process, nodeIds) => {
+    return _.some(this.getAllGroups(process), (group) => this.groupIncludesAllOfNodes(group, nodeIds))
+  }
```
Comment (PPiedel) [inline ui/client/components/graph/NodeUtils.js]: These curly braces are unnecessary ?

File: ui/client/actions/notificationActions.js [hunk 5]
```diff
@@ -21,4 +21,13 @@ export function error(message, error, showErrorText) {
...
     ])
   })
+}
+
+export function info(message) {
+  return Notifications.info({
+    autoDismiss: 10,
+    children: ([
+      <Notification icon={InlinedSvgs.tipsInfo} message={message} key={uuid4()}/>
+    ]),
```
Comment (PPiedel) [inline ui/client/actions/notificationActions.js]: comma

File: ui/client/common/ProcessUtils.js [hunk 6]
```diff
@@ -155,7 +155,7 @@ class ProcessUtils {
 
   //TODO: this should be done without these switches..
   findNodeDefinitionId = (node) => {
-    switch (node.type) {
+    switch (_.get(node, "type")) {
```
Comment (PPiedel) [inline ui/client/common/ProcessUtils.js]: I'm wondering if we really need to change this call into lodash method? We do not pass default value, so what are the benefits?

Comment (mateuszkp96) [reply ui/client/common/ProcessUtils.js]: Sometimes this method is called with undefined as node. This causes an error described in #511.  
With lodash get there is no need to check if object has type property

File: ui/client/components/graph/NodeUtils.js (hunk 4 above)
Comment (mateuszkp96) [reply ui/client/components/graph/NodeUtils.js]: removed

File: ui/client/actions/notificationActions.js (hunk 5 above)
Comment (mateuszkp96) [reply ui/client/actions/notificationActions.js]: removed

**Pull request merged**